src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx

```diff
@@ -49,76 +149,47 @@ export const useLessonEngineContext = () => {
   return context;
 };
 
-export const LessonEngineProvider = memo((props: { children: ReactNode }) => {
-  // consolidate into a single stateful object
-  const [currentSection, setCurrentSection] =
-    useState<LessonSection>("overview");
-
-  const [completedSections, setCompletedSections] = useState<LessonSection[]>(
-    [],
-  );
-
-  const [sectionResults, setSectionResults] = useState<LessonSectionResults>(
-    {},
-  );
-
-  const getIsComplete = (section: LessonSection) =>
-    completedSections.includes(section);
-
-  const completeSection = (section: LessonSection) => {
-    // concatenate and dedupe the array
-    setCompletedSections((prev) => {
-      const _completedSections = [...prev, section].filter(
-        (item, index, array) => array.indexOf(item) === index,
-      );
-
-      // redirect the user according to what sections have been completed
-      if (
-        _completedSections.length ===
-        lessonSections.filter((s) => s !== "overview" && s !== "review").length
-      ) {
-        setCurrentSection("review");
-      } else {
-        setCurrentSection("overview");
-      }
+type LessonEngineProviderProps = {
+  children: ReactNode;
+  initialLessonReviewSections: Readonly<LessonReviewSection[]>;
+};
 
-      return _completedSections;
+export const LessonEngineProvider = memo(
+  ({ children, initialLessonReviewSections }: LessonEngineProviderProps) => {
+    const [state, dispatch] = useReducer(lessonEngineReducer, {
+      lessonReviewSections: initialLessonReviewSections,
+      currentSection: "overview",
+      sections: {},
     });
-  };
-
-  const updateCurrentSection = (section: LessonSection) => {
-    setCurrentSection(section);
-  };
+    const completeSection = (section: LessonReviewSection) => {
+      dispatch({ type: "completeSection", section });
+    };
+    const updateCurrentSection = (section: LessonSection) =>
+      dispatch({ type: "setCurrentSection", section });
+    const proceedToNextSection = () =>
+      dispatch({ type: "proceedToNextSection" });
+    const updateQuizResult = (result: QuizResult) => {
+      dispatch({ type: "updateQuizResult", result });
+    };
```

Comment:
I'd probably ditch these wrappers and expose `dispatch` directly to reduce boilerplate — however for now it keeps the change minimal